### ISSUE TITLE
Wrong code generation in core library with ROPI-RWPI relocation model on ARMv7em.

### ISSUE THREAD

**Issue by Acciente717** [rust-lang/rust#95871]
### TL; DR version:
When targeting `thumbv7em-none-eabi` with `ropi-rwpi` relocation model, some generated instructions in `core::fmt::num` are wrong. The below naive example won't work in the **debug** build.
```rust
#[no_mangle]
pub extern "C" fn main() -> ! {
    // It prints out "2;", not "256".
    hprintln!("{}", 256);  // ARM semihosting print.
    loop {}
}
```

To reproduce, see the repo below:
https://github.com/Acciente717/ropi-rwpi-err-demo

### Detailed version:
Tracking down with GDB, the wrong instructions are in `_ZN4core3fmt3num3imp7fmt_u32#hash`. The corresponding rust source code is [here](https://github.com/rust-lang/rust/blob/3ea44938e21f0de8ae7d4f6399a8a30f97867c70/library/core/src/fmt/num.rs#L211-L274).

The specific Rust statement leading to wrong instructions is:
```rust
static DEC_DIGITS_LUT: &[u8; 200] = b"0001020304050607080910111213141516171819\
      2021222324252627282930313233343536373839\
      4041424344454647484950515253545556575859\
      6061626364656667686970717273747576777879\
      8081828384858687888990919293949596979899";

let lut_ptr = DEC_DIGITS_LUT.as_ptr();
```

The wrong instructions are:
```
805b322:  f243 7124   movw  r1, #14116    ; 0x3724
805b326:  f2c0 0101   movt  r1, #1
805b32a:  4648        mov   r0, r9
805b32c:  5840        ldr   r0, [r0, r1]
805b32e:  21c8        movs  r1, #200    ; 0xc8
805b330:  f003 fc29   bl    805eb86 <_ZN4core5slice29_$LT$impl$u20$$u5b$T$u5d$$GT$6as_ptr17h58e80ef4ecfdc5b4E>
```

#### Why is it wrong?
In ROPI-RWPI relocation model, read-only data should be stored in Flash, and is accessed by PC-relative addressing. Read-write data should be stored in SRAM, and is accessed by R9-relative addressing. R9 is reserved for this purpose in RWPI, and is called the "static base".

`DEC_DIGITS_LUT` is read-only, but the generated code uses R9-relative addressing. It is essentially accessing some wild memory.

#### Do you have a minimal example?
Unfortunately, no. As can be seen in the GitHub repo linked above, I tried to pull out the function from `core` and compile it alone. But this time, the compiler gives out the correct instructions:
```
80007d2:  f241 504a   movw  r0, #5450    ; 0x154a
80007d6:  f2c0 0006   movt  r0, #6
80007da:  4478        add   r0, pc
80007dc:  6800        ldr   r0, [r0, #0]
80007de:  21c8        movs  r1, #200    ; 0xc8
80007e0:  f05e f9d1   bl    805eb86 <_ZN4core5slice29_$LT$impl$u20$$u5b$T$u5d$$GT$6as_ptr17h58e80ef4ecfdc5b4E>
```

I have no idea how to fix the bug by myself, but I would like to participate in diagnosing and get it fixed as soon as possible.

### Meta

`rustc --version --verbose`:
```
rustc 1.62.0-nightly (f4a7ce997 2022-04-08)
binary: rustc
commit-hash: f4a7ce997a1d7546d2b737f8b87d36907bcea2ad
commit-date: 2022-04-08
host: x86_64-unknown-linux-gnu
release: 1.62.0-nightly
LLVM version: 14.0.0
```

**Comment by Urgau** [comment]
*Given that the problem seems to be in the assembly code generated I'm tagging this as an LLVM related issue but feel free to remove the LLVM label if you disagree.*

@rustbot label +I-prioritize +A-llvm +T-compiler

**Comment by Acciente717** [comment]
Just wanted to follow up, is there anything I can help to speed up the fix?

**Comment by jacobbramley** [comment]
I've no real hypothesis yet, but some observations that may be helpful:
* I get the same `r9`-based code generation for `thumbv7m-none-eabi` (Cortex-M3), so it appears to be possible to hit something similar on the LM3S6965, with appropriate edits to `memory.x` and so on. (That's handy because stock Qemu supports it.)
* The failure I see (on LM3S6965-Qemu) is actually different — `from_utf8` fails and the subsequent `unwrap()` sends execution to the `HardFault` handler — but _I think_ it is caused by a similar issue, reading the wrong memory somewhere.

At this stage, I'm not sure what would help to speed it up, assuming that you can't reduce the test case further :-)

**Comment by Acciente717** [comment]
Thank you for your investigation. I will also try to minimize the example.

**Comment by jacobbramley** [comment]
I have a couple of findings this morning:
* A simpler type doesn't work because the compiler inlines the constant. Ok, not a big problem.
* More usefully, I've found the actual trigger for this: putting the function inside a module.

That reduces it to a very simple, single-file test case:

In a brand new project (`cargo new --lib ...`):

#### `src/lib.rs`
```rust
#![no_std]
static VAR: &[u8] = b"42";

pub fn test() -> *const u8 {
    VAR.as_ptr()
}

pub mod nested {
    pub fn test() -> *const u8 {
        super::VAR.as_ptr()
    }
}
```

#### `.cargo/config.toml`
```toml
[target.thumbv7m-none-eabi]
rustflags = [
  "-C", "relocation-model=rwpi"
]

[build]
target = "thumbv7m-none-eabi"
```

`test()` compiles correctly, but `nested::test()` uses `r9` to load the address of `VAR`, as in `fmt_u32`.

The LLVM IR is identical for the two `test` functions, except for some `!dbg` annotations. I will investigate that a bit further.

**Comment by Acciente717** [comment]
That's an incredible finding. I successfully reproduced the minimal buggy example in my environment. Neither did I find anything different in the LLVM IR except `!dbg`.

**Comment by jacobbramley** [comment]
This reproduces with the following `rustc` command:
```
rustc lib.rs --crate-name repro --crate-type lib --emit=link --target=thumbv7m-none-eabi -C relocation-model=rwpi
```
That generates `librepro.rlib`, with the `r9` bug.

What's odd is that emitting anything else (other than `metadata` or `dep-info`) makes the bug disappear _from every output_ (including `librepro.rlib`). Even `--emit=obj,link` is sufficient to get correct code generation for this test (in both the `.o` and the `.rlib`). Those flags don't seem to be as independent as I'd assumed, and I suspect that's why we don't see it in `asm` or `llvm-ir` output.

In the short term, you can probably work around your bug by adding `,asm` to your `--emit` argument, but there's surely a bug in here that still needs investigation.

**Comment by apiraino** [comment]
Assigning priority as discussed in the [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2395871.20Wrong.20code.20generation.20in.20core.20library.20with.20ROPI-RWPI.E2.80.A6) of the Prioritization Working Group.

@rustbot label -I-prioritize +P-medium

**Comment by vjonaswolf** [comment]
I just filed a PR for LLVM to fix this issue.
